{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "authentic-kernel",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scib\n",
    "import scanpy as sc\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "southwest-wagon",
   "metadata": {},
   "source": [
    "The data was taken from here https://figshare.com/articles/dataset/Benchmarking_atlas-level_data_integration_in_single-cell_genomics_-_integration_task_datasets_Immune_and_pancreas_/12420968"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "extraordinary-holly",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2023-03-14 20:51:52--  https://figshare.com/ndownloader/files/24539942\n",
      "Resolving figshare.com (figshare.com)... 54.229.56.157, 34.240.113.119, 2a05:d018:1f4:d000:8f2d:b761:e885:37ef, ...\n",
      "Connecting to figshare.com (figshare.com)|54.229.56.157|:443... connected.\n",
      "HTTP request sent, awaiting response... 302 Found\n",
      "Location: https://s3-eu-west-1.amazonaws.com/pfigshare-u-files/24539942/Lung_atlas_public.h5ad?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIYCQYOYV5JSSROOA/20230314/eu-west-1/s3/aws4_request&X-Amz-Date=20230314T205152Z&X-Amz-Expires=10&X-Amz-SignedHeaders=host&X-Amz-Signature=7507b3705a5322e5902dbf68ec5b9593bdbfdf33a8b2605d6eddce4234b00498 [following]\n",
      "--2023-03-14 20:51:52--  https://s3-eu-west-1.amazonaws.com/pfigshare-u-files/24539942/Lung_atlas_public.h5ad?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIYCQYOYV5JSSROOA/20230314/eu-west-1/s3/aws4_request&X-Amz-Date=20230314T205152Z&X-Amz-Expires=10&X-Amz-SignedHeaders=host&X-Amz-Signature=7507b3705a5322e5902dbf68ec5b9593bdbfdf33a8b2605d6eddce4234b00498\n",
      "Resolving s3-eu-west-1.amazonaws.com (s3-eu-west-1.amazonaws.com)... 52.218.24.2, 52.218.45.192, 52.218.101.51, ...\n",
      "Connecting to s3-eu-west-1.amazonaws.com (s3-eu-west-1.amazonaws.com)|52.218.24.2|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 1019664664 (972M) [application/octet-stream]\n",
      "Saving to: ‘data/Lung_atlas_public.h5ad’\n",
      "\n",
      "data/Lung_atlas_pub 100%[===================>] 972.43M  92.1MB/s    in 10s     \n",
      "\n",
      "2023-03-14 20:52:03 (93.1 MB/s) - ‘data/Lung_atlas_public.h5ad’ saved [1019664664/1019664664]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!wget -O ../data/Lung_atlas_public.h5ad https://figshare.com/ndownloader/files/24539942"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "accepting-association",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jovyan/my-conda-envs/scnew/lib/python3.10/site-packages/anndata/_io/h5ad.py:238: OldFormatWarning: Element '/layers' was written without encoding metadata.\n",
      "  d[k] = read_elem(f[k])\n",
      "/home/jovyan/my-conda-envs/scnew/lib/python3.10/site-packages/anndata/_io/specs/methods.py:584: OldFormatWarning: Element '/obs/__categories/dataset' was written without encoding metadata.\n",
      "  categories = read_elem(categories_dset)\n",
      "/home/jovyan/my-conda-envs/scnew/lib/python3.10/site-packages/anndata/_io/specs/methods.py:587: OldFormatWarning: Element '/obs/dataset' was written without encoding metadata.\n",
      "  read_elem(dataset), categories, ordered=ordered\n",
      "/home/jovyan/my-conda-envs/scnew/lib/python3.10/site-packages/anndata/_io/specs/methods.py:584: OldFormatWarning: Element '/obs/__categories/location' was written without encoding metadata.\n",
      "  categories = read_elem(categories_dset)\n",
      "/home/jovyan/my-conda-envs/scnew/lib/python3.10/site-packages/anndata/_io/specs/methods.py:587: OldFormatWarning: Element '/obs/location' was written without encoding metadata.\n",
      "  read_elem(dataset), categories, ordered=ordered\n",
      "/home/jovyan/my-conda-envs/scnew/lib/python3.10/site-packages/anndata/_io/specs/methods.py:590: OldFormatWarning: Element '/obs/nGene' was written without encoding metadata.\n",
      "  return read_elem(dataset)\n",
      "/home/jovyan/my-conda-envs/scnew/lib/python3.10/site-packages/anndata/_io/specs/methods.py:590: OldFormatWarning: Element '/obs/nUMI' was written without encoding metadata.\n",
      "  return read_elem(dataset)\n",
      "/home/jovyan/my-conda-envs/scnew/lib/python3.10/site-packages/anndata/_io/specs/methods.py:584: OldFormatWarning: Element '/obs/__categories/patientGroup' was written without encoding metadata.\n",
      "  categories = read_elem(categories_dset)\n",
      "/home/jovyan/my-conda-envs/scnew/lib/python3.10/site-packages/anndata/_io/specs/methods.py:587: OldFormatWarning: Element '/obs/patientGroup' was written without encoding metadata.\n",
      "  read_elem(dataset), categories, ordered=ordered\n",
      "/home/jovyan/my-conda-envs/scnew/lib/python3.10/site-packages/anndata/_io/specs/methods.py:590: OldFormatWarning: Element '/obs/percent.mito' was written without encoding metadata.\n",
      "  return read_elem(dataset)\n",
      "/home/jovyan/my-conda-envs/scnew/lib/python3.10/site-packages/anndata/_io/specs/methods.py:584: OldFormatWarning: Element '/obs/__categories/protocol' was written without encoding metadata.\n",
      "  categories = read_elem(categories_dset)\n",
      "/home/jovyan/my-conda-envs/scnew/lib/python3.10/site-packages/anndata/_io/specs/methods.py:587: OldFormatWarning: Element '/obs/protocol' was written without encoding metadata.\n",
      "  read_elem(dataset), categories, ordered=ordered\n",
      "/home/jovyan/my-conda-envs/scnew/lib/python3.10/site-packages/anndata/_io/specs/methods.py:584: OldFormatWarning: Element '/obs/__categories/sanger_type' was written without encoding metadata.\n",
      "  categories = read_elem(categories_dset)\n",
      "/home/jovyan/my-conda-envs/scnew/lib/python3.10/site-packages/anndata/_io/specs/methods.py:587: OldFormatWarning: Element '/obs/sanger_type' was written without encoding metadata.\n",
      "  read_elem(dataset), categories, ordered=ordered\n",
      "/home/jovyan/my-conda-envs/scnew/lib/python3.10/site-packages/anndata/_io/specs/methods.py:590: OldFormatWarning: Element '/obs/size_factors' was written without encoding metadata.\n",
      "  return read_elem(dataset)\n",
      "/home/jovyan/my-conda-envs/scnew/lib/python3.10/site-packages/anndata/_io/specs/methods.py:584: OldFormatWarning: Element '/obs/__categories/sampling_method' was written without encoding metadata.\n",
      "  categories = read_elem(categories_dset)\n",
      "/home/jovyan/my-conda-envs/scnew/lib/python3.10/site-packages/anndata/_io/specs/methods.py:587: OldFormatWarning: Element '/obs/sampling_method' was written without encoding metadata.\n",
      "  read_elem(dataset), categories, ordered=ordered\n",
      "/home/jovyan/my-conda-envs/scnew/lib/python3.10/site-packages/anndata/_io/specs/methods.py:584: OldFormatWarning: Element '/obs/__categories/batch' was written without encoding metadata.\n",
      "  categories = read_elem(categories_dset)\n",
      "/home/jovyan/my-conda-envs/scnew/lib/python3.10/site-packages/anndata/_io/specs/methods.py:587: OldFormatWarning: Element '/obs/batch' was written without encoding metadata.\n",
      "  read_elem(dataset), categories, ordered=ordered\n",
      "/home/jovyan/my-conda-envs/scnew/lib/python3.10/site-packages/anndata/_io/specs/methods.py:584: OldFormatWarning: Element '/obs/__categories/cell_type' was written without encoding metadata.\n",
      "  categories = read_elem(categories_dset)\n",
      "/home/jovyan/my-conda-envs/scnew/lib/python3.10/site-packages/anndata/_io/specs/methods.py:587: OldFormatWarning: Element '/obs/cell_type' was written without encoding metadata.\n",
      "  read_elem(dataset), categories, ordered=ordered\n",
      "/home/jovyan/my-conda-envs/scnew/lib/python3.10/site-packages/anndata/_io/specs/methods.py:584: OldFormatWarning: Element '/obs/__categories/donor' was written without encoding metadata.\n",
      "  categories = read_elem(categories_dset)\n",
      "/home/jovyan/my-conda-envs/scnew/lib/python3.10/site-packages/anndata/_io/specs/methods.py:587: OldFormatWarning: Element '/obs/donor' was written without encoding metadata.\n",
      "  read_elem(dataset), categories, ordered=ordered\n",
      "/home/jovyan/my-conda-envs/scnew/lib/python3.10/site-packages/anndata/_io/specs/methods.py:590: OldFormatWarning: Element '/obs/index' was written without encoding metadata.\n",
      "  return read_elem(dataset)\n",
      "/home/jovyan/my-conda-envs/scnew/lib/python3.10/site-packages/anndata/_io/specs/methods.py:590: OldFormatWarning: Element '/var/index' was written without encoding metadata.\n",
      "  return read_elem(dataset)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 32472 × 15148\n",
       "    obs: 'dataset', 'location', 'nGene', 'nUMI', 'patientGroup', 'percent.mito', 'protocol', 'sanger_type', 'size_factors', 'sampling_method', 'batch', 'cell_type', 'donor'\n",
       "    layers: 'counts'"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata = sc.read('../data/Lung_atlas_public.h5ad')\n",
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "stuffed-humor",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([1.2873256, 1.2873256, 1.2873256, ..., 0.8000515, 1.2165804,\n",
       "       0.7779638], dtype=float32)"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.X.data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "brown-singles",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([1.       , 1.       , 1.       , ..., 1.1387753, 2.2072291,\n",
       "       1.0936011], dtype=float32)"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.layers['counts'].data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "statutory-needle",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dataset</th>\n",
       "      <th>location</th>\n",
       "      <th>nGene</th>\n",
       "      <th>nUMI</th>\n",
       "      <th>patientGroup</th>\n",
       "      <th>percent.mito</th>\n",
       "      <th>protocol</th>\n",
       "      <th>sanger_type</th>\n",
       "      <th>size_factors</th>\n",
       "      <th>sampling_method</th>\n",
       "      <th>batch</th>\n",
       "      <th>cell_type</th>\n",
       "      <th>donor</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>index</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GAGGGGCCAGCC</th>\n",
       "      <td>Dropseq_transplant</td>\n",
       "      <td>Dropseq_lower</td>\n",
       "      <td>249</td>\n",
       "      <td>355</td>\n",
       "      <td>nan</td>\n",
       "      <td>0.019718</td>\n",
       "      <td>drop-seq</td>\n",
       "      <td>drop-seq</td>\n",
       "      <td>0.381231</td>\n",
       "      <td>Transplant</td>\n",
       "      <td>B1</td>\n",
       "      <td>Type 2</td>\n",
       "      <td>B1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACATAGATAGGC</th>\n",
       "      <td>Dropseq_transplant</td>\n",
       "      <td>Dropseq_lower</td>\n",
       "      <td>582</td>\n",
       "      <td>987</td>\n",
       "      <td>nan</td>\n",
       "      <td>0.044580</td>\n",
       "      <td>drop-seq</td>\n",
       "      <td>drop-seq</td>\n",
       "      <td>1.081643</td>\n",
       "      <td>Transplant</td>\n",
       "      <td>B1</td>\n",
       "      <td>Type 2</td>\n",
       "      <td>B1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ATCGACGCGGCA</th>\n",
       "      <td>Dropseq_transplant</td>\n",
       "      <td>Dropseq_lower</td>\n",
       "      <td>292</td>\n",
       "      <td>490</td>\n",
       "      <td>nan</td>\n",
       "      <td>0.026531</td>\n",
       "      <td>drop-seq</td>\n",
       "      <td>drop-seq</td>\n",
       "      <td>0.497352</td>\n",
       "      <td>Transplant</td>\n",
       "      <td>B1</td>\n",
       "      <td>Type 2</td>\n",
       "      <td>B1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CCAGCTTAGGTT</th>\n",
       "      <td>Dropseq_transplant</td>\n",
       "      <td>Dropseq_lower</td>\n",
       "      <td>1057</td>\n",
       "      <td>1806</td>\n",
       "      <td>nan</td>\n",
       "      <td>0.035437</td>\n",
       "      <td>drop-seq</td>\n",
       "      <td>drop-seq</td>\n",
       "      <td>2.172617</td>\n",
       "      <td>Transplant</td>\n",
       "      <td>B1</td>\n",
       "      <td>Type 2</td>\n",
       "      <td>B1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCATAGAGTCCT</th>\n",
       "      <td>Dropseq_transplant</td>\n",
       "      <td>Dropseq_lower</td>\n",
       "      <td>716</td>\n",
       "      <td>1378</td>\n",
       "      <td>nan</td>\n",
       "      <td>0.020319</td>\n",
       "      <td>drop-seq</td>\n",
       "      <td>drop-seq</td>\n",
       "      <td>1.197578</td>\n",
       "      <td>Transplant</td>\n",
       "      <td>B1</td>\n",
       "      <td>Type 2</td>\n",
       "      <td>B1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LungTranscriptome7239219_LungTranscriptome7239219GGGCATCCACCAACCG</th>\n",
       "      <td>10x_Transplant</td>\n",
       "      <td>Transplant</td>\n",
       "      <td>1178</td>\n",
       "      <td>3245</td>\n",
       "      <td>Parenchyma</td>\n",
       "      <td>0.011896</td>\n",
       "      <td>10x v2</td>\n",
       "      <td>other</td>\n",
       "      <td>0.775650</td>\n",
       "      <td>Transplant</td>\n",
       "      <td>5</td>\n",
       "      <td>Mast cell</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LungTranscriptome7239219_LungTranscriptome7239219TGCTGCTAGCAATCTC</th>\n",
       "      <td>10x_Transplant</td>\n",
       "      <td>Transplant</td>\n",
       "      <td>1919</td>\n",
       "      <td>5984</td>\n",
       "      <td>Parenchyma</td>\n",
       "      <td>0.018747</td>\n",
       "      <td>10x v2</td>\n",
       "      <td>other</td>\n",
       "      <td>0.982236</td>\n",
       "      <td>Transplant</td>\n",
       "      <td>5</td>\n",
       "      <td>Mast cell</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LungTranscriptome7239213_GACTGCGTCCTGCCAT</th>\n",
       "      <td>10x_Transplant</td>\n",
       "      <td>Transplant</td>\n",
       "      <td>4411</td>\n",
       "      <td>25233</td>\n",
       "      <td>Parenchyma</td>\n",
       "      <td>0.010489</td>\n",
       "      <td>10x v2</td>\n",
       "      <td>other</td>\n",
       "      <td>1.235763</td>\n",
       "      <td>Transplant</td>\n",
       "      <td>3</td>\n",
       "      <td>Mast cell</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LungTranscriptome7239218_GCTGCGACATTCACTT</th>\n",
       "      <td>10x_Transplant</td>\n",
       "      <td>Transplant</td>\n",
       "      <td>2132</td>\n",
       "      <td>7216</td>\n",
       "      <td>Parenchyma</td>\n",
       "      <td>0.013827</td>\n",
       "      <td>10x v2</td>\n",
       "      <td>other</td>\n",
       "      <td>1.025105</td>\n",
       "      <td>Transplant</td>\n",
       "      <td>6</td>\n",
       "      <td>Mast cell</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LungTranscriptome7239219_LungTranscriptome7239219TTAGGACGTGACGGTA</th>\n",
       "      <td>10x_Transplant</td>\n",
       "      <td>Transplant</td>\n",
       "      <td>1886</td>\n",
       "      <td>5846</td>\n",
       "      <td>Parenchyma</td>\n",
       "      <td>0.017164</td>\n",
       "      <td>10x v2</td>\n",
       "      <td>other</td>\n",
       "      <td>0.929115</td>\n",
       "      <td>Transplant</td>\n",
       "      <td>5</td>\n",
       "      <td>Mast cell</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>32472 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                               dataset  \\\n",
       "index                                                                    \n",
       "GAGGGGCCAGCC                                        Dropseq_transplant   \n",
       "ACATAGATAGGC                                        Dropseq_transplant   \n",
       "ATCGACGCGGCA                                        Dropseq_transplant   \n",
       "CCAGCTTAGGTT                                        Dropseq_transplant   \n",
       "TCATAGAGTCCT                                        Dropseq_transplant   \n",
       "...                                                                ...   \n",
       "LungTranscriptome7239219_LungTranscriptome72392...      10x_Transplant   \n",
       "LungTranscriptome7239219_LungTranscriptome72392...      10x_Transplant   \n",
       "LungTranscriptome7239213_GACTGCGTCCTGCCAT               10x_Transplant   \n",
       "LungTranscriptome7239218_GCTGCGACATTCACTT               10x_Transplant   \n",
       "LungTranscriptome7239219_LungTranscriptome72392...      10x_Transplant   \n",
       "\n",
       "                                                         location  nGene  \\\n",
       "index                                                                      \n",
       "GAGGGGCCAGCC                                        Dropseq_lower    249   \n",
       "ACATAGATAGGC                                        Dropseq_lower    582   \n",
       "ATCGACGCGGCA                                        Dropseq_lower    292   \n",
       "CCAGCTTAGGTT                                        Dropseq_lower   1057   \n",
       "TCATAGAGTCCT                                        Dropseq_lower    716   \n",
       "...                                                           ...    ...   \n",
       "LungTranscriptome7239219_LungTranscriptome72392...     Transplant   1178   \n",
       "LungTranscriptome7239219_LungTranscriptome72392...     Transplant   1919   \n",
       "LungTranscriptome7239213_GACTGCGTCCTGCCAT              Transplant   4411   \n",
       "LungTranscriptome7239218_GCTGCGACATTCACTT              Transplant   2132   \n",
       "LungTranscriptome7239219_LungTranscriptome72392...     Transplant   1886   \n",
       "\n",
       "                                                     nUMI patientGroup  \\\n",
       "index                                                                    \n",
       "GAGGGGCCAGCC                                          355          nan   \n",
       "ACATAGATAGGC                                          987          nan   \n",
       "ATCGACGCGGCA                                          490          nan   \n",
       "CCAGCTTAGGTT                                         1806          nan   \n",
       "TCATAGAGTCCT                                         1378          nan   \n",
       "...                                                   ...          ...   \n",
       "LungTranscriptome7239219_LungTranscriptome72392...   3245   Parenchyma   \n",
       "LungTranscriptome7239219_LungTranscriptome72392...   5984   Parenchyma   \n",
       "LungTranscriptome7239213_GACTGCGTCCTGCCAT           25233   Parenchyma   \n",
       "LungTranscriptome7239218_GCTGCGACATTCACTT            7216   Parenchyma   \n",
       "LungTranscriptome7239219_LungTranscriptome72392...   5846   Parenchyma   \n",
       "\n",
       "                                                    percent.mito  protocol  \\\n",
       "index                                                                        \n",
       "GAGGGGCCAGCC                                            0.019718  drop-seq   \n",
       "ACATAGATAGGC                                            0.044580  drop-seq   \n",
       "ATCGACGCGGCA                                            0.026531  drop-seq   \n",
       "CCAGCTTAGGTT                                            0.035437  drop-seq   \n",
       "TCATAGAGTCCT                                            0.020319  drop-seq   \n",
       "...                                                          ...       ...   \n",
       "LungTranscriptome7239219_LungTranscriptome72392...      0.011896    10x v2   \n",
       "LungTranscriptome7239219_LungTranscriptome72392...      0.018747    10x v2   \n",
       "LungTranscriptome7239213_GACTGCGTCCTGCCAT               0.010489    10x v2   \n",
       "LungTranscriptome7239218_GCTGCGACATTCACTT               0.013827    10x v2   \n",
       "LungTranscriptome7239219_LungTranscriptome72392...      0.017164    10x v2   \n",
       "\n",
       "                                                   sanger_type  size_factors  \\\n",
       "index                                                                          \n",
       "GAGGGGCCAGCC                                          drop-seq      0.381231   \n",
       "ACATAGATAGGC                                          drop-seq      1.081643   \n",
       "ATCGACGCGGCA                                          drop-seq      0.497352   \n",
       "CCAGCTTAGGTT                                          drop-seq      2.172617   \n",
       "TCATAGAGTCCT                                          drop-seq      1.197578   \n",
       "...                                                        ...           ...   \n",
       "LungTranscriptome7239219_LungTranscriptome72392...       other      0.775650   \n",
       "LungTranscriptome7239219_LungTranscriptome72392...       other      0.982236   \n",
       "LungTranscriptome7239213_GACTGCGTCCTGCCAT                other      1.235763   \n",
       "LungTranscriptome7239218_GCTGCGACATTCACTT                other      1.025105   \n",
       "LungTranscriptome7239219_LungTranscriptome72392...       other      0.929115   \n",
       "\n",
       "                                                   sampling_method batch  \\\n",
       "index                                                                      \n",
       "GAGGGGCCAGCC                                            Transplant    B1   \n",
       "ACATAGATAGGC                                            Transplant    B1   \n",
       "ATCGACGCGGCA                                            Transplant    B1   \n",
       "CCAGCTTAGGTT                                            Transplant    B1   \n",
       "TCATAGAGTCCT                                            Transplant    B1   \n",
       "...                                                            ...   ...   \n",
       "LungTranscriptome7239219_LungTranscriptome72392...      Transplant     5   \n",
       "LungTranscriptome7239219_LungTranscriptome72392...      Transplant     5   \n",
       "LungTranscriptome7239213_GACTGCGTCCTGCCAT               Transplant     3   \n",
       "LungTranscriptome7239218_GCTGCGACATTCACTT               Transplant     6   \n",
       "LungTranscriptome7239219_LungTranscriptome72392...      Transplant     5   \n",
       "\n",
       "                                                    cell_type donor  \n",
       "index                                                                \n",
       "GAGGGGCCAGCC                                           Type 2    B1  \n",
       "ACATAGATAGGC                                           Type 2    B1  \n",
       "ATCGACGCGGCA                                           Type 2    B1  \n",
       "CCAGCTTAGGTT                                           Type 2    B1  \n",
       "TCATAGAGTCCT                                           Type 2    B1  \n",
       "...                                                       ...   ...  \n",
       "LungTranscriptome7239219_LungTranscriptome72392...  Mast cell     5  \n",
       "LungTranscriptome7239219_LungTranscriptome72392...  Mast cell     5  \n",
       "LungTranscriptome7239213_GACTGCGTCCTGCCAT           Mast cell     3  \n",
       "LungTranscriptome7239218_GCTGCGACATTCACTT           Mast cell     6  \n",
       "LungTranscriptome7239219_LungTranscriptome72392...  Mast cell     5  \n",
       "\n",
       "[32472 rows x 13 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "coupled-snake",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>index</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A1BG-AS1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M-AS1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A4GALT</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAAS</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZXDC</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYG11B</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYX</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZEF1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZZ3</th>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>15148 rows × 0 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: []\n",
       "Index: [A1BG-AS1, A2M, A2M-AS1, A4GALT, AAAS, AACS, AADAC, AADAT, AAED1, AAGAB, AAK1, AAMDC, AAMP, AAR2, AARD, AARS, AARS2, AARSD1, AASDH, AASDHPPT, AASS, AATF, AATK, ABAT, ABCA1, ABCA10, ABCA12, ABCA13, ABCA2, ABCA3, ABCA4, ABCA5, ABCA6, ABCA7, ABCA8, ABCA9, ABCB1, ABCB10, ABCB4, ABCB5, ABCB6, ABCB7, ABCB8, ABCB9, ABCC1, ABCC10, ABCC11, ABCC2, ABCC3, ABCC4, ABCC5, ABCC6, ABCC9, ABCD1, ABCD2, ABCD3, ABCD4, ABCE1, ABCF1, ABCF2, ABCF3, ABCG1, ABCG2, ABHD10, ABHD11, ABHD12, ABHD12B, ABHD13, ABHD14A, ABHD14B, ABHD15, ABHD16A, ABHD17A, ABHD17B, ABHD17C, ABHD2, ABHD3, ABHD4, ABHD5, ABHD6, ABHD8, ABI1, ABI2, ABI3, ABI3BP, ABL1, ABL2, ABLIM1, ABLIM3, ABR, ABRACL, ABT1, ABTB1, AC002064.4, AC002117.1, AC002310.12, AC002310.7, AC002398.11, AC002398.12, AC002451.3, ...]\n",
       "\n",
       "[15148 rows x 0 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.var"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "scnew",
   "language": "python",
   "name": "scnew"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
